docs/architecture/adr-024-coin-metadata.md

```diff
@@ -0,0 +1,145 @@
+application.
+
+With regards to a balance `amount`, the Cosmos SDK assumes a static and fixed unit of denomination,
+regardless of the denomination itself. In other words, clients and apps built atop a Cosmos-SDK-based
+chain may choose to define and use arbitrary units of denomination to provide a richer UX, however, by
+the time a tx or operation reaches the Cosmos SDK state machine, the `amount` is treated as a single
+unit. For example, for the Cosmos Hub (Gaia), clients assume 1 ATOM = 10^6 uatom, and so all txs and
+operations in the Cosmos SDK work off of units of 10^6.
+
+This clearly provides a poor and limited UX especially as interoperability of networks increases and
+as a result the total amount of asset types increases. We propose to have `x/bank` additionally keep
+track of metadata per `denom` in order to help clients, wallet providers, and explorers improve their
+UX and remove the requirement for making any assumptions on the unit of denomination.
+
+## Decision
+
+The `x/bank` module will be updated to store and index metadata by `denom`, specifically the "base" or
+smallest unit -- the unit the Cosmos SDK state-machine works with. The metadata should ideally be
+extensible and serve the needs of many external consumers and even other modules (e.g. `x/ibc` or `x/nft`).
+At a minimum, the metadata type should include all units and a potential list of denomination aliases.
+
+To provide a means of extensibility, metadata may include a non-zero length list of `extensions`,
+which are simply key/value pairs. In addition, metadata may also include a non-zero length list of
+`aliases` with regard to the origin chain (e.g. `["ETH", "ether"]`). Both `extensions` and `aliases`
+may be changed via governance.
+
+As a result, we can define the type as follows:
+
+```protobuf
+message DenomUnit {
+  string denom    = 1;
+  uint32 exponent = 2;  
+  repeated string aliases = 3;
+}
+
+message Extension {
+  key string = 1;
+  value string = 2;
+}
+
+message Metadata {
+  string description = 1;
+  string base = 2;
+  repeated DenomUnit denom_units = 3;
+  repeated string aliases = 4;
+  repeated Extension extensions = 5;
+}
+```
+
+As an example, the ATOM's metadata can be defined as follows:
+
+```json
+{
+  "description": "The native staking token of the Cosmos Hub.",
+  "base": "uatom",
+  "denom_units": [
+    {
+      "denom": "matom",
+      "exponent": 3,
+      "aliases": [
+        "milliatom"
+      ]
+    },
+    {
+      "denom": "atom",
+      "exponent": 6,
+    }
+  ],
+  "aliases": [
+    "microatom"
+  ],
+  "extensions": [
+    {
+      "key": "total_supply",
+      "value": "543578243452349991"
+    },
+    {
+      "key": "max_supply",
+      "value": "999999999999999999"
+    },
+    {
+      "key": "source_chain",
+      "value": "gaia"
+    }
+  ]
+}
+```
+
+> Given the above metadata, a client may infer that 4.3atom = 4.3 * (10^6) = 4,300,000uatom.
```

Comment:
Wouldn't the other way round be much more interesting?

> Given a balance/fee/amount of `4,300,000uatom` on chain, how to display that to the user in the most meaningful way?

It would be hard to answer that from the structure above.

Reply:
I don't know, would it? 

> Given a balance/fee/amount/supply of 4,300,000uatom on chain, how to display that to the user in the most meaningful way?

4,300,000/(10^6) = 4.3atom

Reply:
> I don't know, would it?

For the use cases I am aware of, it would. See e.g. https://github.com/CosmWasm/code-explorer/blob/6f5618d/src/ui-utils/index.ts#L27-L30 to display a `Coin`. It makes two assumtions which would be great to have on-chain:
1. the prefix "u" means micro
2. a token name is to be capitalized

Given the data in this document and no external information, the client does not know which of those to display:
- 4.3atom
- 4300matom
- 4300000uatom

Similarly, the ticker symbol ATOM that every user interface uses it not included in here. It is needed for input boxes like
<img width="593" alt="Bildschirmfoto 2020-05-29 um 22 28 30" src="https://user-images.githubusercontent.com/2603011/83302509-c445e680-a1fb-11ea-8af0-fa37fecc6426.png">

---------------

I did not follow the motivation for the multiple aliases. But from my perspective you only need the raw version (integer units) and the user visible version (symbol+exponent). Like

```csv
uatom,ATOM,6
wei,ETH,18
niov,IOV,9
house,HOUSE,0
```

Reply:
A client can display whatever is wishes. It knows the base unit (e.g. uatom) and then it knows all the units. If an account has a balance `"4300000"`, then the client can convert this to any one of those units.

Reply:
> A client can display whatever is wishes.

But isn't the whole point of this document to remove the need of hardcoded currency information? A client can only "wish" something it understands. But given the on-chain data, how would the client decide? Pick the first option? Pick random? Build a fancy algorithm that uses the shortest representation (which then gets inconsistent across the UI)?

I think the minimal addition we'd need here would be to declare one of those options as the "default" or "display" option. For ATOM it would be `{ "denom": "atom", "exponent": 6, }`.

Reply:
> But isn't the whole point of this document to remove the need of hardcoded currency information? 

No, it's not the motivation of this ADR.

> But given the on-chain data, how would the client decide? Pick the first option? Pick random?

No -- all a chain needs to know is (1) the denom and (2) the base unit. From there, it can convert at will.

Reply:
Okay, then I guess I misunderstood

> This clearly provides a poor and limited UX especially as interoperability of networks increases and
as a result the total amount of asset types increases. We propose to have `x/bank` additionally keep
track of metadata per `denom` in order to help clients, wallet providers, and explorers improve their
UX and remove the requirement for making any assumptions on the unit of denomination.

and this document has a different motivation I'm not aware of. Is it to allow different input denoms in transactions that are converted to the base automatically before being stored? For this direction the design makes sense. But I don't see why reading clients like explorers would benefit from it. Wallets could potentially offer multiple denom options in an UI, but as far as I can see, only one is used in real life.

For the record, with such a change almost every client still needs the following information locally for every supported coin:
1. Primary display unit or ticker symbol (e.g. "ATOM")
2. Exponent to base (e.g. 6)

Reply:
The original document in https://github.com/cosmos/cosmos-sdk/pull/2072/files with `1 [Base Unit] = 10^(N) [Smallest Unit]` being available on-chain would solve all client side needs I am aware of.

Reply:
> Is it to allow different input denoms in transactions that are converted to the base automatically before being stored?

No, as the ADR states the state-machine assumes only a single unit (e.g. `uatom`). This feature is solely to allow clients to understand the units better and to allow APIs exist to convert between units.

Reply:
Why was this discussion resolved? You want to provide token meta data for the client to interprete the coin with on-chain data but in the current state you cannot even do the most simple task: display an amount in the prepresentation the user expects.

Reply:
> display an amount in the prepresentation the user expects.

I'm sorry, I don't follow.

1. You know the base denom (e.g. `uatom`)
2. You can query for all units using this denom
3. You can display any other representation of this token using the metadata (e.g. atom)
4. The ledger still only deals (consumes and returns) the base denom

The metadata is solely used for clients to know what to display. And btw, this doesn't even have to belong the ledger -- it's just a convenience. So what is the issue exactly?